classes/core/PKPPageRouter.php

```diff
@@ -495,6 +514,75 @@ private function _getRequestedUrlParts($callback, $request): array|string|null
         return $callback($url ?? '', $userVars);
     }
+
+    /**
+     * Get context object and context/site/all locales.
+     */
+    private function _getContextAndLocales(PKPRequest $request, string $contextPath): array
+    {
+        return [
+            $context = $this->getCurrentContext() ?? (($contextPath === 'index' || !$contextPath || $contextPath === Application::CONTEXT_ID_ALL)
+                ? null
+                : Application::getContextDAO()->getByPath($contextPath)),
+            $context?->getSupportedLocales()
+                ?? (($contextPath === 'index')
+                    ? (Application::isInstalled()
+                        ? $request->getSite()->getSupportedLocales()
+                        : array_keys(Locale::getLocales()))
+                    : [])
+        ];
+    }
+
+    /**
+     * Get locale for URL from session or primary
+     */
+    private function _getLocaleForUrl(PKPRequest $request, ?Context $context, array $locales, ?string $urlLocaleForPage): string
+    {
+        return in_array($locale = $urlLocaleForPage ?: Locale::getLocale(), $locales)
+            ? $locale
+            : (($context ?? $request->getSite())?->getPrimaryLocale() ?? Locale::getLocale());
+    }
+
+    /**
+     * Change the locale for the current user.
+     * Redirect to url with(out) locale if locale changed or context set to multi/monolingual.
+     */
+    private function _setLocale(PKPRequest $request, ?string $setLocale): void
+    {
+        $contextPath = $this->_getRequestedUrlParts(['Core', 'getContextPath'], $request);
+        $urlLocale = $this->_getRequestedUrlParts(['Core', 'getLocalization'], $request);
+        $multiLingual = count($this->_getContextAndLocales($request, $contextPath)[1]) > 1;
+
+        if (!$multiLingual && !$urlLocale && !$setLocale || $multiLingual && !$setLocale && $urlLocale === Locale::getLocale()) {
+            return;
+        }
+
+        $sessionLocale = (function (string $l) use ($request): string {
+            $session = $request->getSession();
+            if (Locale::isSupported($l) && $l !== $session->get('currentLocale')) {
+                $session->put('currentLocale', $l);
+                $request->setCookieVar('currentLocale', $l);
+            }
+            // In case session current locale has been set to non-supported locale, or is null, somewhere else
+            if (!Locale::isSupported($session->get('currentLocale') ?? "")) {
+                $session->put('currentLocale', Locale::getLocale());
+                $request->setCookieVar('currentLocale', Locale::getLocale());
+            }
+            return $session->get('currentLocale');
+        })($setLocale ?? $urlLocale);
+
+        if (preg_match('#^/\w#', $source = $request->getUserVar('source') ?? "")) {
+            $request->redirectUrl($source);
+        }
+
+        $uri = Core::removeBaseUrl($setLocale ? ($_SERVER['HTTP_REFERER'] ?? "") : ($_SERVER['REQUEST_URI'] ?? ""));
```

Comment:
Hi @jyhein, in order to support the case that Alec mentioned, I think this line should be something like:
```
        $uri = $_SERVER['PATH_INFO'];
        if ($setLocale) {
            $uri = Core::removeBaseUrl($_SERVER['HTTP_REFERER'] ?? "");
            if (!empty($request->getBasePath())) {
                $uri = str_replace($request->getBasePath(), '', $uri);
            }
        }

```
Maybe we should also somehow test if it also works if the journal is using/has configured the following part of the config.inc.php:
```

; Base URL override settings: Entries like the following examples can
; be used to override the base URLs used by OJS. If you want to use a
; proxy to rewrite URLs to OJS, configure your proxy's URL with this format.
; Syntax: base_url[journal_path] = http://www.example.com
;
; Example1: URLs that aren't part of a particular journal.
;    Example1: base_url[index] = http://www.example.com
; Example2: URLs that map to a subdirectory.
;    Example2: base_url[myJournal] = http://www.example.com/myJournal
; Example3: URLs that map to a subdomain.
;    Example3: base_url[myOtherJournal] = http://myOtherJournal.example.com
 
```
I think you have such installations, no?